internal/servers/controller/session_cleanup_job.go

```diff
@@ -0,0 +1,191 @@
+package controller
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/scheduler"
+	"github.com/hashicorp/boundary/internal/servers"
+	"github.com/hashicorp/boundary/internal/servers/controller/common"
+	"github.com/hashicorp/go-hclog"
+)
+
+// sessionCleanupJob defines a periodic job that monitors workers for
+// loss of connection and terminates connections on workers that have
+// not sent a heartbeat in a significant period of time.
+//
+// Relevant connections are simply marked as disconnected in the
+// database. Connections will be independently terminated by the
+// worker, or the event of a synchronization issue between the two,
+// the controller will win out and order that the connections be
+// closed on the worker.
+type sessionCleanupJob struct {
+	logger        hclog.Logger
+	serversRepoFn common.ServersRepoFactory
+	sessionRepoFn common.SessionRepoFactory
+
+	// The amount of time to give disconnected workers before marking
+	// their connections as closed.
+	statusGracePeriod time.Duration
+
+	// Worker count, which is probably the only visible metric we will
+	// be working with. Actual connections closed will be a record
+	// count metric that we will probably only log, if anything.
+	numProcessed, numWorkers int
+}
+
+// newSessionCleanupJob instantiates the session cleanup job.
+func newSessionCleanupJob(
+	logger hclog.Logger,
+	serversRepoFn common.ServersRepoFactory,
+	sessionRepoFn common.SessionRepoFactory,
+	statusGracePeriod time.Duration,
+) (*sessionCleanupJob, error) {
+	const op = "controller.newNewSessionCleanupJob"
+	switch {
+	case logger == nil:
+		return nil, errors.New(errors.InvalidParameter, op, "missing logger")
+	case serversRepoFn == nil:
+		return nil, errors.New(errors.InvalidParameter, op, "missing serversRepoFn")
+	case sessionRepoFn == nil:
+		return nil, errors.New(errors.InvalidParameter, op, "missing sessionRepoFn")
+	}
+
+	return &sessionCleanupJob{
+		logger:            logger,
+		serversRepoFn:     serversRepoFn,
+		sessionRepoFn:     sessionRepoFn,
+		statusGracePeriod: statusGracePeriod,
+	}, nil
+}
+
+// Name returns a short, unique name for the job.
+func (j *sessionCleanupJob) Name() string { return "sessionCleanupJob" }
+
+// Description returns the description for the job.
+func (j *sessionCleanupJob) Description() string {
+	return "Clean up session connections from disconnected workers"
+}
+
+// NextRunIn returns the next run time after a job is completed.
+//
+// The next run time is defined for sessionCleanupJob as one second.
+// This is because the job should run continuously to terminate
+// connections as soon as a worker has not reported in for a long
+// enough time. Only one job will ever run at once, so there is no
+// reason why it cannot run again immediately.
+func (j *sessionCleanupJob) NextRunIn() (time.Duration, error) { return time.Second, nil }
```

Comment:
While I don't necessarily disagree with the comment a second does seem aggressive.  Regardless of that please note that each scheduler only requests work once a minute, so you will only truly reach continuously running at scale when multiple controllers are running schedulers.

Reply:
Yeah, I discovered that after setting this up, hah.

One thing that I'm worried about is that if I set this, and we add a next run time of say, like a minute, that it will possibly be two minutes before the job runs again, as the next scheduled time may not come up before the next job run happens.

Is there any other/better way to guarantee that this job runs once per minute? The actual grace period is 15 seconds, so we want to make absolute sure that it runs as soon as possible to ensure connections are closed as soon as they can be.

Reply:
Discussed this with @louisruch offline and agreed on the fact that the 1 second interval should be good for now to ensure that the job always runs within a particular controller's minimum scheduler interval (1 minute). This will be increasingly variable as the number of controllers go up (and assuming that they were _not_ all started at around the same time), but it helps cover cases where there will be a small amount of controllers where the possibility of the actual next run time being a full minute is higher.